### ISSUE TITLE
ShouldAllBeEquivalentTo and Excluding is not working properly

### ISSUE THREAD

**Issue by OzgurDogan** [dennisdoomen/fluentassertions#194]
When you use the extension ShouldAllBeEquivalentTo on enumerable and also want to exclude a property then it's not working properly.

Example:

        public void Test()
        {
            var list1 = new[] { new KeyValuePair<int, int>(1, 123) };
            var list2 = new[] { new KeyValuePair<int, int>(2, 321) };

            list1.ShouldAllBeEquivalentTo(list2, config => config.Excluding(ctx => ctx.Key));
        }

The example should fail because the value is also different, but it doesn't

When you use ShouldBeEquivalentTo then it fails the test like expected

        public void Test()
        {
            var list1 = new[] { new KeyValuePair<int, int>(1, 123) };
            var list2 = new[] { new KeyValuePair<int, int>(2, 321) };

            list1[0].ShouldBeEquivalentTo(list2[0], config => config.Excluding(ctx => ctx.Key));
        }

**Comment by vossad01** [comment]
Found the issue and have a fix.  Want to spend some time to make sure the change won't introduce any more issues.  I am a bit ashamed that I broke this first and this issue revealed not one but two errors in my code.  This regression was first released in 3.1.229.

@dennisdoomen I don't know what the current release schedule is like for `develop`. I can prepare a PR against `master` as well if you would like because this was a regression.

**Comment by dennisdoomen** [comment]
@vossad01 I was hoping to release something before the end of January. I'm still working on the global configuration thing though. When that's done, and we have done some dogfooding, v3.4 should be a fact.

**Comment by thrandre** [comment]
Are there still issues with this? Almost all our tests that are using ShouldAllBeEquivalentTo are failing. We are using the latest release (3.3.0). We have assertions like this (just an example):

```
fagkoder.ShouldAllBeEquivalentTo(new List<GrepFagkode> {fagkodeOne, fagkodeTwo}, o => o.ExcludingProperties().ExcludingNestedObjects().ExcludingFields().ExcludingMissingMembers());
```

This fails with the current result:

```
Expected item[0].Id to be 0, but found 3.

Expected item[0].Kandidattyper[0].Id to be 0, but found 3.

Expected item[0].Kandidattyper[0].FagkodeId to be 0, but found 3.

Expected item[0].Kandidattyper[0].ShouldBeImported to be 

False, but found 

True.

Expected item[1].Id to be 0, but found 4.

Expected item[1].Kandidattyper[0].Id to be 0, but found 4.

Expected item[1].Kandidattyper[0].FagkodeId to be 0, but found 4.

Expected item[1].Kandidattyper[0].ShouldBeImported to be 

False, but found 

True.


With configuration:
- Use declared types and members

- Compare enums by value

- Include all non-private fields

- Match member by name (or throw)

- Be strict about the order of items in byte arrays

- FluentAssertions.Equivalency.ShouldAllBeEquivalentToHelper+CollectionMemberOrderingRuleDecorator
```

The new global assertion options are working fine though:

```
AssertionOptions.AssertEquivalencyUsing(f => f.ExcludingProperties());
```

**Comment by dennisdoomen** [comment]
What does the type represented by `fagkoder` look like?

**Comment by thrandre** [comment]
It's a ```List<GrepFagkode>```. GrepFagkode is this little beauty:

```
 public class GrepFagkode
    {
        public int Id { get; set; }
		
		public string Kode { get; set; }
	    public DateTime? SistEndret { get; set; }
	    public string OpplaeringsnivaaId { get; set; }
	    public string OppgaveansvarId { get; set; }
	    public string SensuransvarId { get; set; }
	    public string FoersteSemesterId { get; set; }
	    public string SisteSemesterId { get; set; }
	    public bool ShouldBeImported { get; set; }
	    public string ExcludedFromImportReason { get; set; }

	    public List<GrepKandidattype> Kandidattyper { get; set; }
	    public List<GrepTekstOversettelse> Navn { get; set; }

	    public GrepFagkode()
		{
			Navn = new List<GrepTekstOversettelse>();
			Kandidattyper = new List<GrepKandidattype>();
		}
    }
```

**Comment by dennisdoomen** [comment]
So you have a collection of these objects and you compare them to another collection? Is it supposed to exclude any properties?

**Comment by thrandre** [comment]
That is correct. We actually just want to base equivalency on the "Kode" property (using Include(f => f.Kode)), but I found that it doesn't consider the defined options at all. Using ShouldBeEquivalent in a loop works just fine.

**Comment by dennisdoomen** [comment]
Huh? You're saying that `actual.ShouldAllBeEquivalentTo(expected, options => options.Include(f => f.Kode))` is not working for you?

**Comment by thrandre** [comment]
Yep, from what I can see it is trying to match all the other properties as well (I have also tried excluding all properties and then including the single "Kode").

**Comment by dennisdoomen** [comment]
Need to investigate that.

**Comment by thrandre** [comment]
Perfect. Thanks!

**Comment by mzizka** [comment]
Quick question here. I expected `ShouldAllBeEquivalentTo()` to yield the same results as looping over objects and calling `ShouldBeEquivalent()` on each one. My issue, possibly related to this one, is that  is that the `ExcludingNestedObjects()` option seems to have no impact when comparing lists (v3.3.0):

```C#
[TestClass]
public class ExcludingNestedObjectsTests
{
    private class MyChildObject
    {
        public int Id { get; set; }
        public string MyChildString { get; set; }

        public override bool Equals(object obj)
        {
            return obj is MyChildObject
                   && ((MyChildObject)obj).Id == this.Id;
        }
    }

    private class MyObject
    {
        public string MyString { get; set; }
        public MyChildObject Child { get; set; }
    }

    private static MyObject CreateMyObject()
    {
        MyObject object1 = new MyObject { MyString = "identical string" };
        object1.Child = new MyChildObject { Id = 1, MyChildString = "identical string" };
        return object1;
    }

    [TestMethod]
    public void ShouldBeEquivalentTest()
    {
        MyObject object1 = CreateMyObject();
        MyObject object2 = CreateMyObject();
        IList<MyObject> objectList1 = new List<MyObject> { object1 };
        IList<MyObject> objectList2 = new List<MyObject> { object2 };

        object1.ShouldBeEquivalentTo(object2, opt => opt.ExcludingNestedObjects()); // Pass
        object1.ShouldBeEquivalentTo(object2); // Pass

        objectList1.ShouldAllBeEquivalentTo(objectList2, opt => opt.ExcludingNestedObjects()); // Pass
        objectList1.ShouldAllBeEquivalentTo(objectList2); // Pass

        object1.Child.MyChildString = "DIFFERENT STRING";

        // At this point, if we do not exclude nested objects, it will fail.

        // This will pass:
        object1.ShouldBeEquivalentTo(object2, opt => opt.ExcludingNestedObjects());

        // This will fail, unexpectedly:
        objectList1.ShouldAllBeEquivalentTo(objectList2, opt => opt.ExcludingNestedObjects());
    }
}
```

**Comment by dennisdoomen** [comment]
It should be honored. Have to find some time to analyze.

**Comment by dennisdoomen** [comment]
I've just found some time to test this issue with the latest version and I can confirm it still fails.

**Comment by dennisdoomen** [comment]
Found the problem. When you call `ExcludingNestedObjects` you're basically telling FA to only compare the properties of the root object and to never traverse the object graph. For `ShouldBeEquivalentTo` this means that it will only use `Equals` on the properties of `MyObject` and never go into the properties of `MyChildObject`. However, when you use `ShouldAllBeEquivalentTo`, the concept of _root_ is wrong. It will treat the top-level enumerable as the _root_ causing it to do a simple equality check on the two `MyObject` instances. Since that type doesn't override `Object.Equals`, this results is a simple reference equality check.

**Comment by mzizka** [comment]
Thanks !

**Comment by jannavarro** [comment]
@dennisdoomen I have version 4.0 of FluentAssertions and I still see FA comparing nested objects when I do this:

objectList1.ShouldAllBeEquivalentTo(objectList2, opt => opt.ExcludingNestedObjects());

Did I understand correct that this should be on v3.4?

**Comment by dennisdoomen** [comment]
Yes, it *should* have been fixed in 3.4.
